wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,262 @@
+		return netaddr.IP{}, dns.RCodeRefused, errNotOurName
+	}
+
+	r.mu.Lock()
+	if r.dnsMap == nil {
+		r.mu.Unlock()
+		return netaddr.IP{}, dns.RCodeServerFailure, errMapNotSet
+	}
+	addr, found := r.dnsMap.DomainToIP[domain]
+	r.mu.Unlock()
+
+	if !found {
+		return netaddr.IP{}, dns.RCodeNameError, errNoSuchDomain
+	}
+	return addr, dns.RCodeSuccess, nil
+}
+
+type response struct {
+	Header         dns.Header
+	ResourceHeader dns.ResourceHeader
+	Question       dns.Question
+	IP             netaddr.IP
+}
+
+func (r *Resolver) parseQuery(query *packet.ParsedPacket, resp *response) error {
+	var err error
+
+	resp.Header, err = r.parser.Start(query.Payload())
+	if err != nil {
+		resp.Header.RCode = dns.RCodeFormatError
+		return err
+	}
+
+	if resp.Header.Response {
+		resp.Header.RCode = dns.RCodeFormatError
+		return errNotQuery
+	}
+
+	resp.Question, err = r.parser.Question()
+	if err != nil {
+		resp.Header.RCode = dns.RCodeFormatError
+		return err
+	}
+
+	return nil
+}
+
+func (r *Resolver) makeResponse(resp *response) error {
+	var err error
+
+	switch resp.Question.Type {
+	case dns.TypeA, dns.TypeALL:
+		// Remove final dot from name: *.ipn.dev. -> *.ipn.dev
+		name := resp.Question.Name.String()
+		name = name[:len(name)-1]
+		resp.IP, resp.Header.RCode, err = r.Resolve(name)
+	default:
+		resp.Header.RCode = dns.RCodeNotImplemented
+		err = errNotImplemented
+	}
+
+	return err
+}
+
+func marshalAnswer(resp *response, builder *dns.Builder) error {
+	var answer dns.AResource
+
+	err := builder.StartAnswers()
+	if err != nil {
+		return err
+	}
+
+	answerHeader := dns.ResourceHeader{
+		Name:  resp.Question.Name,
+		Type:  dns.TypeA,
+		Class: dns.ClassINET,
+		TTL:   3600,
+	}
+	ip := resp.IP.As16()
+	copy(answer.A[:], ip[12:])
+	return builder.AResource(answerHeader, answer)
+}
+
+// marshalResponse serializes resp by appending
+func marshalResponse(resp *response, out []byte) ([]byte, error) {
+	resp.Header.Response = true
+	resp.Header.Authoritative = true
+	if resp.Header.RecursionDesired {
+		resp.Header.RecursionAvailable = true
+	}
+
+	builder := dns.NewBuilder(out, resp.Header)
+
+	err := builder.StartQuestions()
+	if err != nil {
+		return nil, err
+	}
+
+	err = builder.Question(resp.Question)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Header.RCode == dns.RCodeSuccess {
+		err = marshalAnswer(resp, &builder)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return builder.Finish()
+}
+
+// Respond generates a response to the given packet.
+// It is assumed that r.AcceptsPacket(query) is true.
+func (r *Resolver) Respond(query *packet.ParsedPacket) ([]byte, error) {
+	var resp response
+
+	// 0. Generate response header.
+	udpHeader := query.UDPHeader()
+	udpHeader.ToResponse()
+
+	// 1. Parse query packet.
+	err := r.parseQuery(query, &resp)
+	// We will not return this error: it is the sender's fault.
+	if err != nil {
+		r.logf("tsdns: error during query parsing: %v", err)
+		goto respond
```

Comment:
Using goto like this is not very idiomatic in go, usually I would make all the code after `respond:` into a helper function, and have all these branches finish with `return r.sendResponse(blah)` or something.

I don't feel strongly about this however, I'm fine with submitting this function as-is. Just something to consider.